### PR TITLE
Add Bernoulli distribution

### PR DESCRIPTION
**Pull request by YoshikawaMasashi**
split from #4678.

### CODE REVIEW

**Comment by YoshikawaMasashi** [comment]
jenkins, test this please.

File: chainer/distributions/bernoulli.py [hunk 1]
```diff
@@ -0,0 +1,100 @@
...
+from chainer.backends import cuda
+from chainer import distribution
+from chainer.functions.activation import sigmoid
+from chainer.functions.array import broadcast
+from chainer.functions.math import exponential
+import numpy
+
+
+class Bernoulli(distribution.Distribution):
+
```
Comment (toslunar) [inline chainer/distributions/bernoulli.py]: Could you add a docstring with meaning of arguments `p`, `logits`?

File: chainer/distributions/bernoulli.py [hunk 2]
```diff
@@ -0,0 +1,119 @@
...
+                "Either `p` or `logit` (not both) must have a value.")
+
+        with chainer.using_config('enable_backprop', True):
+            if p is None:
+                self.logit = chainer.as_variable(logit)
+                self.p = sigmoid.sigmoid(self.logit)
+            else:
+                self.p = chainer.as_variable(p)
+                self.logit = exponential.log(self.p) - \
+                    exponential.log(1. - self.p)
```
Comment (toslunar) [inline chainer/distributions/bernoulli.py]: To avoid casting, `log1p(-self.p)` can be used.  Maybe you'd like to cache `p.dtype.type(1) - p` in the both cases.

File: chainer/distributions/bernoulli.py [hunk 3]
```diff
@@ -0,0 +1,119 @@
...
+
+    @property
+    def batch_shape(self):
+        return self.p.shape
+
+    @property
+    def entropy(self):
+        return -self.p * exponential.log(self.p) - \
```
Comment (toslunar) [inline chainer/distributions/bernoulli.py]: While `x * log(x)` converges to `0` as x → +0, the result is `0 * inf = nan`.

File: chainer/distributions/bernoulli.py [hunk 4]
```diff
@@ -0,0 +1,119 @@
...
+            (numpy.float32(1.) - self.p) \
```
Comment (toslunar) [inline chainer/distributions/bernoulli.py]: How about the following?
```
p = self.p
q = p.dtype.type(1.) - p
```

File: chainer/distributions/bernoulli.py [hunk 5]
```diff
@@ -0,0 +1,119 @@
...
+    @property
+    def event_shape(self):
+        return ()
+
+    @property
+    def _is_gpu(self):
+        return isinstance(self.p.data, cuda.ndarray)
+
+    def log_prob(self, x):
+        if type(x) == chainer.Variable:
```
Comment (toslunar) [inline chainer/distributions/bernoulli.py]: Use `isinstance` to support subclasses (e.g. `chainer.Parameter`).

File: chainer/distributions/bernoulli.py [hunk 6]
```diff
@@ -0,0 +1,119 @@
...
+            x = x.data
```
Comment (toslunar) [inline chainer/distributions/bernoulli.py]: Since `distribution` is new, `.array` is preferred to `.data`.

File: chainer/distributions/bernoulli.py [hunk 7]
```diff
@@ -0,0 +1,119 @@
...
+    def stddev(self):
+        return (self.p * (1 - self.p)) ** 0.5
+
+    @property
+    def support(self):
+        return '{0, 1}'
+
+    @property
+    def variance(self):
+        return (self.p * (1 - self.p))
```
Comment (toslunar) [inline chainer/distributions/bernoulli.py]: The outer parentheses is unnecessary.

File: tests/chainer_tests/distributions_tests/test_bernoulli.py [hunk 8]
```diff
@@ -0,0 +1,39 @@
+from chainer import distributions
+from chainer import testing
+import numpy
+
+
+@testing.parameterize(*testing.product({
+    'shape': [(3, 2), (1,)],
+    'is_variable': [True, False],
+    'sample_shape': [(3, 2), ()],
+}))
```
Comment (toslunar) [inline tests/chainer_tests/distributions_tests/test_bernoulli.py]: Add a boolean parameter to test with `p` that contains `0` and `1`.
(You don't have to add it to the KL divergence test.)

File: chainer/distributions/bernoulli.py [hunk 9]
```diff
@@ -0,0 +1,119 @@
...
+    def _is_gpu(self):
+        return isinstance(self.p.data, cuda.ndarray)
+
+    def log_prob(self, x):
+        if type(x) == chainer.Variable:
+            x = x.data
+        x = x.astype(self.p.dtype)
+        if self._is_gpu:
+            invalid_inf = cuda.cupy.zeros_like(x)
+            invalid = cuda.cupy.bitwise_and(x != 0, x != 1)
```
Comment (toslunar) [inline chainer/distributions/bernoulli.py]: Why do you want to validate `x`?  Do you expect other integers, or any floats?

File: chainer/distributions/bernoulli.py [hunk 10]
```diff
@@ -0,0 +1,119 @@
...
+
+    @property
+    def batch_shape(self):
+        return self.p.shape
+
+    @property
+    def entropy(self):
+        return -self.p * exponential.log(self.p) - \
+            (numpy.float32(1.) - self.p) \
+            * exponential.log(numpy.float32(1.) - self.p)
```
Comment (toslunar) [inline chainer/distributions/bernoulli.py]: Be consistent in breaking a line after/before binary operator (pycodestyle's W503 or W504).  I prefer linebreaks before binary operators (as this line).

File: chainer/distributions/bernoulli.py (hunk 3 above)
Comment (YoshikawaMasashi) [reply chainer/distributions/bernoulli.py]: Oh...
As a matter of fact, when `p=0`, entropy is `-q * log(q) = 0`.
I'll fix. (but when `p=0`, entropy is not differentiable.)

File: chainer/distributions/bernoulli.py (hunk 9 above)
Comment (YoshikawaMasashi) [reply chainer/distributions/bernoulli.py]: Yes. When x is not 0 or 1, this function returns -inf.

File: tests/chainer_tests/distributions_tests/test_bernoulli.py (hunk 8 above)
Comment (YoshikawaMasashi) [reply tests/chainer_tests/distributions_tests/test_bernoulli.py]: For now, I don't expect `p` whose type is `bool`, and we get error when boolean `p` is used (Because `F.log` does not support bool.).
And scipy.stats also don't support boolean `p`.
I think we don't have to support boolen `p`. How about this?

Comment (toslunar) [reply tests/chainer_tests/distributions_tests/test_bernoulli.py]: What I tried to say is
```
if self.p_has_extreme_values:
    some_indices = ...
    p[some_indices] = np.random.randint(2, size=...)
```

File: chainer/distributions/bernoulli.py (hunk 9 above)
Comment (YoshikawaMasashi) [reply chainer/distributions/bernoulli.py]: Um.., I also think that it is better to support an option not to validate as like `F.bernoulli_nll`.

File: chainer/distributions/bernoulli.py [hunk 11]
```diff
@@ -0,0 +1,186 @@
...
+        return gy * dx,
+
+
+def _bernoulli_log_prob(logit, x):
+    y, = BernoulliLogProb().apply((logit, x))
+    return y
+
+
+def _modified_xlogx(x):
+    return ModifiedXLogX(exponential.log(x)).apply((x,))[0]
```
[the remark this answers is not in the document]
Comment (YoshikawaMasashi) [reply chainer/distributions/bernoulli.py]: Oh, `where` is useful..

**Comment by YoshikawaMasashi** [comment]
jenkins, test this please.

File: chainer/distributions/bernoulli.py [hunk 12]
```diff
@@ -59,27 +56,40 @@ def __init__(self, logx):
...
     return y
 
 
 def _modified_xlogx(x):
-    return ModifiedXLogX(exponential.log(x)).apply((x,))[0]
+    xp = cuda.get_array_module(x)
+    if x.ndim == 0:
+        x = reshape.reshape(x, (1,))
+    if isinstance(x, chainer.Variable):
+        return ModifiedXLogX(exponential.log(
```
Comment (toslunar) [inline chainer/distributions/bernoulli.py]: Do not repeat the application of `ModifiedXLogX`. How about storing logx, or
```
x = chainer.as_variable(x)
xp = x.xp
```

Comment (YoshikawaMasashi) [reply chainer/distributions/bernoulli.py]: Oh yes!

**Comment by YoshikawaMasashi** [comment]
> Could you explain why reshape is needed if ndim == 0?

```
return ModifiedXLogX(exponential.log(
            where.where(x.array > 0, x, xp.ones_like(x.array)))).apply((x,))[0]
```
When `x.ndim == 0`, `x.array > 0` become scalar. And `where` does not work.

**Comment by toslunar** [comment]
How about `utils.force_array(x.array > 0)`?

**Comment by YoshikawaMasashi** [comment]
fixed. thanks.

**Comment by YoshikawaMasashi** [comment]
jenkins, test this please.

**Comment by YoshikawaMasashi** [comment]
jenkins, test this please.
